File: frontend/src/services/weights.ts
/**
 * 权重库相关API服务
 *
 * 提供权重文件的上传、查询、删除等功能
 */

import { apiClient } from './api';

// ==============================
// 类型定义
// ==============================

export type TaskType = 'classification' | 'detection' | 'auto';

export type Framework = 'pytorch' | 'onnx';

// ==============================
// 训练配置详情相关类型
// ==============================

export interface DatasetInfo {
  id: number;
  name: string;
  format: string;
  num_images?: number;
  num_classes?: number;
  classes?: string[];
  path?: string;
}

export interface ModelArchitectureInfo {
  id: number;
  name: string;
  description?: string;
  file_path?: string;
  input_size?: number[];
  task_type?: string;
}

export interface PretrainedWeightInfo {
  id: number;
  name: string;
  display_name: string;
  task_type: string;
  version: string;
  source_type: 'uploaded' | 'trained';
}

export interface AugmentationConfigInfo {
  enabled: boolean;
  strategy?: string;
  strategy_id?: number;
  description?: string;
  config?: Record<string, any> | any[];  // 支持对象和数组（pipeline）
}

export interface WeightUploadData {
  file: File;
  name: string;
  task_type: TaskType;
  description?: string;
  input_size?: number[];
  class_names?: string[];
  normalize_params?: Record<string, any>;
}

export interface WeightLibraryItem {
  id: number;
  name: string;
  display_name: string;
  description?: string;
  task_type: TaskType;
  version: string;
  file_name: string;
  file_size_mb?: number;
  framework: Framework;
  is_auto_detected: boolean;
  is_root?: boolean;
  source_type?: 'uploaded' | 'trained';
  architecture_id?: number;
  parent_version_id?: number;
  created_at: string;
}

export interface WeightLibraryListResponse {
  weights: WeightLibraryItem[];
  total: number;
}

export interface WeightUploadResponse {
  success: boolean;
  message: string;
  weight_id: number;
  weight: WeightLibraryItem;
}

export interface WeightUploadOptions {
  name: string;
  task_type: TaskType;
  description?: string;
  input_size?: number[];
  class_names?: string[];
  normalize_params?: Record<string, any>;
}

// ==============================
// 权重树形结构相关类型
// ==============================

export interface WeightTreeItem extends WeightLibraryItem {
  is_root: boolean;
  source_type: 'uploaded' | 'trained';
  source_training_id?: number;
  architecture_id?: number;
  children: WeightTreeItem[];
}

export interface WeightRootListResponse {
  weights: WeightLibraryItem[];
  total: number;
}

export interface WeightTrainingConfig {
  weight_id: number;
  weight_name: string;
  training_config: Record<string, any> | null;
  source_training: {
    id: number;
    name: string;
    status?: string;
    created_at?: string;
    hyperparams?: Record<string, any>;
  } | null;
  // 完整的关联信息
  dataset?: DatasetInfo | null;
  model_architecture?: ModelArchitectureInfo | null;
  pretrained_weight?: PretrainedWeightInfo | null;
  augmentation?: AugmentationConfigInfo | null;
}

export interface WeightForTrainingOption {
  id: number;
  name: string;
  display_name: string;
  description?: string;
  task_type: TaskType;
  version: string;
  file_path: string;
  architecture_id?: number;
  architecture_name?: string;
  created_at: string;
}

// ==============================
// API方法
// ==============================

class WeightService {
  /**
   * 上传权重文件
   * @param data 上传数据
   * @param onProgress 上传进度回调
   */
  async uploadWeight(
    data: WeightUploadData,
    onProgress?: (progress: number) => void
  ): Promise<WeightUploadResponse> {
    const formData = new FormData();
    formData.append('file', data.file);
    formData.append('name', data.name);
    formData.append('task_type', data.task_type);
    if (data.description) {
      formData.append('description', data.description);
    }
    if (data.input_size) {
      formData.append('input_size', JSON.stringify(data.input_size));
    }
    if (data.class_names) {
      formData.append('class_names', JSON.stringify(data.class_names));
    }
    if (data.normalize_params) {
      formData.append('normalize_params', JSON.stringify(data.normalize_params));
    }

    const token = localStorage.getItem('access_token');
    const baseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1';

    // 使用原生fetch以支持文件上传进度
    return new Promise<WeightUploadResponse>((resolve, reject) => {
      const xhr = new XMLHttpRequest();

      // 监听上传进度
      xhr.upload.addEventListener('progress', (e) => {
        if (e.lengthComputable && onProgress) {
          const progress = Math.round((e.loaded / e.total) * 100);
          onProgress(progress);
        }
      });

      // 监听完成
      xhr.addEventListener('load', () => {
        if (xhr.status >= 200 && xhr.status < 300) {
          try {
            const result: WeightUploadResponse = JSON.parse(xhr.responseText);
            resolve(result);
          } catch (e) {
            reject(new Error('解析响应失败'));
          }
        } else {
          try {
            const error = JSON.parse(xhr.responseText);
            const errorMsg = error.detail || error.message || '上传失败';
            reject(new Error(errorMsg));
          } catch {
            reject(new Error(`上传失败 (${xhr.status}): ${xhr.statusText}`));
          }
        }
      });

      // 监听错误
      xhr.addEventListener('error', () => {
        reject(new Error('网络错误，请检查连接'));
      });

      // 监听取消
      xhr.addEventListener('abort', () => {
        reject(new Error('上传已取消'));
      });

      // 发送请求
      xhr.open('POST', `${baseUrl}/weights/upload`);
      if (token) {
        xhr.setRequestHeader('Authorization', `Bearer ${token}`);
      }
      xhr.send(formData);
    });
  }

  /**
   * 获取权重列表
   */
  async getWeights(taskType?: TaskType): Promise<WeightLibraryListResponse> {
    const params = taskType ? { task_type: taskType } : undefined;
    return await apiClient.get<WeightLibraryListResponse>('/weights', params);
  }

  /**
   * 获取权重详情
   */
  async getWeight(weightId: number): Promise<WeightLibraryItem> {
    return await apiClient.get<WeightLibraryItem>(`/weights/${weightId}`);
  }

  /**
   * 删除权重
   */
  async deleteWeight(weightId: number): Promise<{ message: string; id: number }> {
    return await apiClient.delete<{ message: string; id: number }>(`/weights/${weightId}`);
  }

  /**
   * 创建权重新版本
   */
  async createWeightVersion(
    weightId: number,
    file: File,
    description?: string
  ): Promise<WeightLibraryItem> {
    const formData = new FormData();
    formData.append('file', file);
    if (description) {
      formData.append('description', description);
    }

    const token = localStorage.getItem('access_token');
    const baseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1';

    return new Promise<WeightLibraryItem>((resolve, reject) => {
      const xhr = new XMLHttpRequest();

      xhr.addEventListener('load', () => {
        if (xhr.status >= 200 && xhr.status < 300) {
          try {
            const result: WeightLibraryItem = JSON.parse(xhr.responseText);
            resolve(result);
          } catch (e) {
            reject(new Error('解析响应失败'));
          }
        } else {
          try {
            const error = JSON.parse(xhr.responseText);
            const errorMsg = error.detail || error.message || '创建版本失败';
            reject(new Error(errorMsg));
          } catch {
            reject(new Error(`创建版本失败 (${xhr.status}): ${xhr.statusText}`));
          }
        }
      });

      xhr.addEventListener('error', () => {
        reject(new Error('网络错误，请检查连接'));
      });

      xhr.open('POST', `${baseUrl}/weights/${weightId}/versions`);
      if (token) {
        xhr.setRequestHeader('Authorization', `Bearer ${token}`);
      }
      xhr.send(formData);
    });
  }

  /**
   * 获取权重版本历史
   */
  async getWeightVersions(weightId: number): Promise<WeightLibraryListResponse> {
    return await apiClient.get<WeightLibraryListResponse>(`/weights/${weightId}/versions`);
  }

  /**
   * 按任务类型获取权重列表
   */
  async getWeightsByTask(taskType: TaskType): Promise<WeightLibraryListResponse> {
    return await apiClient.get<WeightLibraryListResponse>(`/weights/by-task/${taskType}`);
  }

  // ==============================
  // 权重树形结构相关API方法
  // ==============================

  /**
   * 获取根节点权重列表
   */
  async getRootWeights(taskType?: TaskType): Promise<WeightRootListResponse> {
    const params = taskType ? { task_type: taskType } : undefined;
    return await apiClient.get<WeightRootListResponse>('/weights/roots', params);
  }

  /**
   * 获取完整的权重树形结构
   */
  async getWeightTree(): Promise<WeightTreeItem[]> {
    return await apiClient.get<WeightTreeItem[]>('/weights/tree');
  }

  /**
   * 获取指定权重的子树
   */
  async getWeightSubtree(weightId: number): Promise<WeightTreeItem> {
    return await apiClient.get<WeightTreeItem>(`/weights/${weightId}/tree`);
  }

  /**
   * 获取权重的训练配置信息
   */
  async getWeightTrainingConfig(weightId: number): Promise<WeightTrainingConfig> {
    return await apiClient.get<WeightTrainingConfig>(`/weights/${weightId}/config`);
  }

  /**
   * 获取可用于训练的权重列表（根据架构筛选）
   */
  async getWeightsForTraining(
    architectureId?: number,
    taskType?: TaskType
  ): Promise<WeightForTrainingOption[]> {
    const params: any = {};
    if (architectureId) params.architecture_id = architectureId;
    if (taskType) params.task_type = taskType;
    return await apiClient.get<WeightForTrainingOption[]>('/weights/for-training', params);
  }

  /**
   * 获取按模型代码筛选的权重树（用于预训练权重选择）
   * 当提供 modelCodeId 时，只返回该模型训练产生的权重
   */
  async getWeightTreeByArchitecture(
    modelCodeId?: number,
    taskType?: TaskType
  ): Promise<WeightTreeItem[]> {
    const params: any = {};
    if (modelCodeId) params.model_code_id = modelCodeId;
    if (taskType) params.task_type = taskType;
    return await apiClient.get<WeightTreeItem[]>('/weights/tree-by-architecture', params);
  }

  /**
   * 按架构ID筛选权重（旧版API，保留兼容）
   * @deprecated 建议使用 getWeightTreeByArchitecture 并传入 modelCodeId
   */
  async getWeightTreeByArchitectureId(
    architectureId: number,
    taskType?: TaskType
  ): Promise<WeightTreeItem[]> {
    const params: any = { architecture_id: architectureId };
    if (taskType) params.task_type = taskType;
    return await apiClient.get<WeightTreeItem[]>('/weights/tree-by-architecture', params);
  }
}

export const weightService = new WeightService();

export default weightService;

File: backend/app/api/v1/weights.py
"""
权重库API路由

提供权重文件的上传、查询、更新、删除等功能。
支持任务类型自动检测和版本管理。

作者: CV Studio 开发团队
日期: 2025-01-15
"""

from fastapi import APIRouter, HTTPException, Query, Depends, UploadFile, File, Form
from typing import List, Optional
from sqlalchemy.orm import Session

from app.services.weight_library_service import WeightLibraryService
from app.schemas.weight_library import (
    WeightLibraryCreate,
    WeightLibraryUpdate,
    WeightLibraryResponse,
    WeightLibraryList,
    WeightLibraryListItem,
    WeightUploadRequest,
    WeightUploadResponse,
    TaskTypeDetectionResponse,
    WeightVersionCreate,
    WeightVersionHistory,
    WeightRootList,
    WeightTreeItem,
    WeightTreeResponse,
    WeightTrainingConfigResponse,
    WeightForTraining,
)
from app.dependencies import get_db
from loguru import logger


router = APIRouter()

# 初始化服务
weight_service = WeightLibraryService()


# ==============================
# API 端点
# ==============================

@router.get("/")
async def weights_root():
    """权重模块根路径"""
    return {
        "message": "权重库模块",
        "status": "运行中",
        "endpoints": {
            "upload": "POST /api/v1/weights/upload - 上传权重文件",
            "list": "GET /api/v1/weights - 获取权重列表",
            "detail": "GET /api/v1/weights/{weight_id} - 获取权重详情",
            "delete": "DELETE /api/v1/weights/{weight_id} - 删除权重",
            "detect": "POST /api/v1/weights/detect - 检测任务类型",
            "versions": "GET /api/v1/weights/{weight_id}/versions - 获取版本历史",
            "create-version": "POST /api/v1/weights/{weight_id}/versions - 创建新版本"
        }
    }


@router.post("/upload", response_model=WeightUploadResponse)
async def upload_weight(
    file: UploadFile = File(..., description="权重文件"),
    name: str = Form(..., description="权重名称"),
    task_type: str = Form(..., description="任务类型：classification/detection/auto"),
    description: Optional[str] = Form(None, description="权重描述"),
    input_size: Optional[str] = Form(None, description="输入尺寸JSON字符串，如 [224, 224]"),
    class_names: Optional[str] = Form(None, description="类别名称JSON字符串，如 [\"cat\", \"dog\"]"),
    normalize_params: Optional[str] = Form(None, description="归一化参数JSON字符串"),
    db: Session = Depends(get_db)
):
    """
    上传权重文件到权重库

    支持的文件格式：.pt, .pth, .pkl, .onnx
    支持的任务类型：classification, detection
    自动检测：task_type设为"auto"时自动检测任务类型

    返回创建的权重记录信息。
    """
    try:
        import json

        # 解析JSON字符串参数
        parsed_input_size = None
        if input_size:
            try:
                parsed_input_size = json.loads(input_size)
            except json.JSONDecodeError:
                raise HTTPException(400, "input_size 格式错误，应为JSON数组")

        parsed_class_names = None
        if class_names:
            try:
                parsed_class_names = json.loads(class_names)
            except json.JSONDecodeError:
                raise HTTPException(400, "class_names 格式错误，应为JSON数组")

        parsed_normalize_params = None
        if normalize_params:
            try:
                parsed_normalize_params = json.loads(normalize_params)
            except json.JSONDecodeError:
                raise HTTPException(400, "normalize_params 格式错误，应为JSON对象")

        # 调用服务层上传
        weight = await weight_service.upload_weight(
            file=file,
            name=name,
            task_type=task_type,
            description=description,
            input_size=parsed_input_size,
            class_names=parsed_class_names,
            normalize_params=parsed_normalize_params,
            uploaded_by=None,  # TODO: 从认证中获取用户ID
            db=db
        )

        return WeightUploadResponse(
            success=True,
            message="权重文件上传成功",
            weight_id=weight.id,
            weight=WeightLibraryResponse.model_validate(weight)
        )

    except ValueError as e:
        raise HTTPException(400, str(e))
    except Exception as e:
        logger.error(f"上传权重失败: {e}")
        raise HTTPException(500, f"上传权重失败: {str(e)}")


@router.get("", response_model=WeightLibraryList)
async def list_weights(
    task_type: Optional[str] = Query(None, description="过滤任务类型"),
    skip: int = Query(0, ge=0, description="跳过数量"),
    limit: int = Query(100, ge=1, le=100, description="返回数量"),
    db: Session = Depends(get_db)
):
    """
    获取权重文件列表

    支持按任务类型过滤和分页。
    """
    try:
        weights = weight_service.get_weights(
            db=db,
            task_type=task_type,
            is_active=True,
            skip=skip,
            limit=limit
        )

        # 统计总数
        total = len(weight_service.get_weights(
            db=db,
            task_type=task_type,
            is_active=True,
            skip=0,
            limit=10000  # 获取全部用于计数
        ))

        items = []
        for weight in weights:
            items.append(WeightLibraryListItem(
                id=weight.id,
                name=weight.name,
                display_name=weight.display_name,
                description=weight.description,
                task_type=weight.task_type,
                version=weight.version,
                file_name=weight.file_name,
                file_size_mb=round(weight.file_size / 1024 / 1024, 2) if weight.file_size else None,
                framework=weight.framework,
                is_auto_detected=weight.is_auto_detected,
                created_at=weight.created_at
            ))

        return WeightLibraryList(
            weights=items,
            total=total
        )

    except Exception as e:
        logger.error(f"获取权重列表失败: {e}")
        raise HTTPException(500, f"获取权重列表失败: {str(e)}")


# ==============================
# 权重树形结构相关端点
# 注意：这些特定路径必须在 /{weight_id} 之前定义！
# ==============================

@router.get("/roots", response_model=WeightRootList)
async def get_root_weights(
    task_type: Optional[str] = Query(None, description="过滤任务类型"),
    skip: int = Query(0, ge=0, description="跳过数量"),
    limit: int = Query(100, ge=1, le=100, description="返回数量"),
    db: Session = Depends(get_db)
):
    """
    获取根节点权重列表

    只返回is_root=True的权重（导入的权重和训练best权重）
    """
    try:
        weights = weight_service.get_root_weights(
            db=db,
            task_type=task_type,
            skip=skip,
            limit=limit
        )

        # 统计总数
        total = len(weight_service.get_root_weights(
            db=db,
            task_type=task_type,
            skip=0,
            limit=10000
        ))

        items = []
        for weight in weights:
            items.append(WeightLibraryListItem(
                id=weight.id,
                name=weight.name,
                display_name=weight.display_name,
                description=weight.description,
                task_type=weight.task_type,
                version=weight.version,
                file_name=weight.file_name,
                file_size_mb=round(weight.file_size / 1024 / 1024, 2) if weight.file_size else None,
                framework=weight.framework,
                is_auto_detected=weight.is_auto_detected,
                is_root=weight.is_root,
                source_type=weight.source_type,
                architecture_id=weight.architecture_id,
                created_at=weight.created_at
            ))

        return WeightRootList(
            weights=items,
            total=total
        )

    except Exception as e:
        logger.error(f"获取根节点权重列表失败: {e}")
        raise HTTPException(500, f"获取根节点权重列表失败: {str(e)}")


@router.get("/tree", response_model=List[WeightTreeResponse])
async def get_weight_tree(
    db: Session = Depends(get_db)
):
    """
    获取完整的权重树形结构

    返回所有根节点及其子节点组成的树
    """
    try:
        trees = weight_service.build_weight_tree(db)
        return trees

    except Exception as e:
        logger.error(f"获取权重树失败: {e}")
        raise HTTPException(500, f"获取权重树失败: {str(e)}")


@router.get("/for-training", response_model=List[WeightForTraining])
async def get_weights_for_training(
    architecture_id: Optional[int] = Query(None, description="模型架构ID"),
    task_type: Optional[str] = Query(None, description="任务类型"),
    db: Session = Depends(get_db)
):
    """
    获取可用于训练的权重列表

    根据模型架构ID和任务类型筛选可用的预训练权重
    """
    try:
        weights = weight_service.get_weights_for_training(
            db=db,
            architecture_id=architecture_id,
            task_type=task_type
        )

        items = []
        for weight in weights:
            # 获取架构名称
            arch_name = None
            if weight.architecture:
                arch_name = weight.architecture.name

            items.append(WeightForTraining(
                id=weight.id,
                name=weight.name,
                display_name=weight.display_name,
                description=weight.description,
                task_type=weight.task_type,
                version=weight.version,
                file_path=weight.file_path,
                architecture_id=weight.architecture_id,
                architecture_name=arch_name,
                created_at=weight.created_at
            ))

        return items

    except Exception as e:
        logger.error(f"获取可用于训练的权重失败: {e}")
        raise HTTPException(500, f"获取可用于训练的权重失败: {str(e)}")


@router.get("/by-task/{task_type}", response_model=WeightLibraryList)
async def list_weights_by_task(
    task_type: str,
    skip: int = Query(0, ge=0, description="跳过数量"),
    limit: int = Query(100, ge=1, le=100, description="返回数量"),
    db: Session = Depends(get_db)
):
    """
    按任务类型获取权重列表

    快捷获取指定任务类型的所有权重。
    """
    try:
        # 验证任务类型
        valid_types = ['classification', 'detection']
        if task_type not in valid_types:
            raise HTTPException(400, f"无效的任务类型，应为: {', '.join(valid_types)}")

        weights = weight_service.get_weights(
            db=db,
            task_type=task_type,
            is_active=True,
            skip=skip,
            limit=limit
        )

        items = []
        for weight in weights:
            items.append(WeightLibraryListItem(
                id=weight.id,
                name=weight.name,
                display_name=weight.display_name,
                description=weight.description,
                task_type=weight.task_type,
                version=weight.version,
                file_name=weight.file_name,
                file_size_mb=round(weight.file_size / 1024 / 1024, 2) if weight.file_size else None,
                framework=weight.framework,
                is_auto_detected=weight.is_auto_detected,
                created_at=weight.created_at
            ))

        return WeightLibraryList(
            weights=items,
            total=len(items)
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"按任务类型获取权重失败: {e}")
        raise HTTPException(500, f"按任务类型获取权重失败: {str(e)}")


# ==============================
# 带参数的权重路由（必须放在最后）
# ==============================

@router.get("/{weight_id}", response_model=WeightLibraryResponse)
async def get_weight(
    weight_id: int,
    db: Session = Depends(get_db)
):
    """
    获取权重文件详情

    根据ID返回权重文件的完整信息。
    """
    try:
        weight = weight_service.get_weight(db, weight_id)
        if not weight:
            raise HTTPException(404, "权重不存在")

        return WeightLibraryResponse.model_validate(weight)

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取权重详情失败: {e}")
        raise HTTPException(500, f"获取权重详情失败: {str(e)}")


@router.delete("/{weight_id}")
async def delete_weight(
    weight_id: int,
    db: Session = Depends(get_db)
):
    """
    删除权重文件

    根据ID删除权重文件（同时删除数据库记录和物理文件）。
    """
    try:
        success = weight_service.delete_weight(db, weight_id)
        if not success:
            raise HTTPException(404, "权重不存在")

        return {"message": "权重已删除", "id": weight_id}

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"删除权重失败: {e}")
        raise HTTPException(500, f"删除权重失败: {str(e)}")


@router.post("/{weight_id}/versions", response_model=WeightLibraryResponse)
async def create_weight_version(
    weight_id: int,
    file: UploadFile = File(..., description="新权重文件"),
    description: Optional[str] = Form(None, description="版本描述"),
    db: Session = Depends(get_db)
):
    """
    创建权重的新版本

    为现有权重创建新版本，自动递增版本号。
    """
    try:
        # 保存上传的文件到临时位置
        import tempfile
        import shutil
        from pathlib import Path

        with tempfile.NamedTemporaryFile(delete=False, suffix=Path(file.filename).suffix) as tmp:
            shutil.copyfileobj(file.file, tmp)
            temp_path = tmp.name

        # 调用服务层创建新版本
        new_weight = weight_service.create_new_version(
            db=db,
            parent_weight_id=weight_id,
            file_path=temp_path,
            description=description
        )

        # 清理临时文件
        Path(temp_path).unlink(missing_ok=True)

        if not new_weight:
            raise HTTPException(404, "父权重不存在")

        return WeightLibraryResponse.model_validate(new_weight)

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"创建权重版本失败: {e}")
        raise HTTPException(500, f"创建权重版本失败: {str(e)}")


@router.get("/{weight_id}/versions", response_model=WeightVersionHistory)
async def get_weight_versions(
    weight_id: int,
    db: Session = Depends(get_db)
):
    """
    获取权重版本历史

    返回指定权重的所有版本信息。
    """
    try:
        versions = weight_service.get_version_history(db, weight_id)

        items = []
        for v in versions:
            items.append(WeightLibraryListItem(
                id=v.id,
                name=v.name,
                display_name=v.display_name,
                description=v.description,
                task_type=v.task_type,
                version=v.version,
                file_name=v.file_name,
                file_size_mb=round(v.file_size / 1024 / 1024, 2) if v.file_size else None,
                framework=v.framework,
                is_auto_detected=v.is_auto_detected,
                created_at=v.created_at
            ))

        return WeightVersionHistory(
            versions=items,
            total=len(items)
        )

    except Exception as e:
        logger.error(f"获取权重版本历史失败: {e}")
        raise HTTPException(500, f"获取权重版本历史失败: {str(e)}")


@router.get("/{weight_id}/tree", response_model=WeightTreeResponse)
async def get_weight_subtree(
    weight_id: int,
    db: Session = Depends(get_db)
):
    """
    获取指定权重的子树

    返回指定权重作为根节点的完整子树
    """
    try:
        subtree = weight_service.get_weight_subtree(db, weight_id)
        if not subtree:
            raise HTTPException(404, "权重不存在")

        return subtree

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取权重子树失败: {e}")
        raise HTTPException(500, f"获取权重子树失败: {str(e)}")


@router.get("/{weight_id}/config", response_model=WeightTrainingConfigResponse)
async def get_weight_training_config(
    weight_id: int,
    db: Session = Depends(get_db)
):
    """
    获取权重的训练配置信息

    返回权重关联的训练配置（如果有）
    """
    try:
        config = weight_service.get_weight_training_config(db, weight_id)
        return config

    except Exception as e:
        logger.error(f"获取权重训练配置失败: {e}")
        raise HTTPException(500, f"获取权重训练配置失败: {str(e)}")
